### DISCUSSION TITLE
deterministic cluster_id values

### DISCUSSION THREAD

**Discussion by aaronamelgar** [moj-analytical-services/splink#873] [Q&A]
If we want to operationalize the Splink, ideally each de-duplicated cluster_id can be re-used in downstream systems. However, we want to re-run the clustering algorithm nightly to detect new duplicates. As I understand it, the cluster_id would be regenerated ever time which prevents us from using it downstream. Is it possible to deterministically assign the cluster_id value?

**Comment by RobinL** [comment]
The cluster ID is the minimum unique_id of nodes in the cluster.

That means that cluster IDs are deterministic so long as clusters are stable. 

If the addition of new records results in changes to clusters you can no longer guarantee they will have the same cluster ID.  However any cluster which is unchanged by the new records will have the same cluster ID. 

Specifically the rule is that the ID of the cluster is equal to the minimum ID of all nodes in the cluster. You can see the code for this [here](https://github.com/moj-analytical-services/splink/blob/aa32ef2ee7fd040fab9574fc6c69fde38aa8d9d4/splink/connected_components.py#L123) and [here](https://github.com/moj-analytical-services/splink/blob/aa32ef2ee7fd040fab9574fc6c69fde38aa8d9d4/splink/connected_components.py#L153)

@ThomasHepworth  is that also your understanding?

**Comment by aaronamelgar** [comment]
thank you @RobinL that does make sense, however I am in the situation you mentioned where new records may result in a new cluster ID. this prevents easily using the cluster ID as a system of record in a downstream process. I would need to create an intermediary process to try and map old cluster ID assigned to new IDs generated. Ideally, I can create some sort of rule based on other columns that generates a custom cluster ID. Open to suggestion on other approaches.

**Reply by RobinL** [reply]
Splink doesn't really solve the problem of incremental clustering for you.  The two solutions that come to mind are either re-running predict() on the full new dataset and reclustering everything, or putting the results from Splink's predict step into a proper graph database like neo4j , and relying on the graph database for the clusters 

In our usage, the way we address this issue is simply reclustering everything when new data comes in.  It's not  obvious to me whether it's actually possible to have deterministic cluster IDs when any new record can merge two existing clusters, but I may be missing something

**Comment by aaronamelgar** [comment]
I 100% agree, rerunning predict on the full dataset is the game plan, which will occur nightly. we re-cluster, but ideally that cluster_id can be reused as our master record id for that person. 

I think essentially we'd need is a way to specify how the cluster_id is constructed rather than always choosing the minimum unique_id. For example, we could use a cleaned version of a blocking rule column of the cluster perhaps the one with highest score, hash it, and then use that as the cluster ID. then any new records that which get clustered in there will be lumped into that email group and therefore generate the same ID.

**Reply by hemanth-reddy-k** [reply]
@aaronamelgar , Were you able to resolve this? I’m currently running into a similar problem, any insights or updates you might have will really help.
